### PR TITLE
List all locks held by a transaction/query

### PR DESCRIPTION
**Pull request by thobe**
Adds a procedure called dbms.listActiveLocks(queryId) that lists all locks held by the transaction that executes the query with the given id.

Sample output from `CALL dbms.listQueries() YIELD queryId CALL dbms.listActiveLocks(queryId) YIELD mode, resourceType, resourceId RETURN *`:
```
+-----------------------------------------------------+
| mode        | queryId   | resourceId | resourceType |
+-----------------------------------------------------+
| "EXCLUSIVE" | "query-1" | 0          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 1          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 2          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 3          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 4          | "NODE"       |
| "SHARED"    | "query-1" | 0          | "SCHEMA"     |
| "SHARED"    | "query-3" | 0          | "SCHEMA"     |
+-----------------------------------------------------+
```

### CODE REVIEW

**Comment by thobe** [comment]
Merging this should probably wait until #8579 has been merged, since it touches the same test files.

**Comment by chrisvest** [comment]
Are they _query ids_ or _transaction ids_?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: I think all the methods that can be accessed concurrently with the running transaction should have javadocs that clearly calls out this fact. The `markForTermination()` method is a good example.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java]: I guess you use this pattern to make the active locks collection use less heap memory?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/locking/ActiveLocksListingCompatibility.java
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/locking/ActiveLocksListingCompatibility.java]: Do we really want to assert on the order? A lock manager implementation may be using hashed sets internally, and requiring an order would force a sort to happen.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/locks/LeaderOnlyLockManager.java
Comment (chrisvest) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/locks/LeaderOnlyLockManager.java]: You should still be able to delegate this to the `localLocks`.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java]: This one should also be able to use the `sharedLocks` and `exclusiveLocks` maps.

File: enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java
Comment (chrisvest) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java:263]: Do we need `activeLocks()` to snapshot its state in a collection, or could it directly return a stream?

Why do we need to transform them to `ActiveLocksQueryResult` objects?

File: enterprise/kernel/src/test/java/org/neo4j/kernel/enterprise/builtinprocs/ListQueriesProcedureTest.java
Comment (chrisvest) [inline enterprise/kernel/src/test/java/org/neo4j/kernel/enterprise/builtinprocs/ListQueriesProcedureTest.java]: It looks odd to me when local variables have all upper case names.

**Review by chrisvest — COMMENTED** [review]
@thobe I had a number of questions and comments that I'd like you to take a look at.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/locking/ActiveLocksListingCompatibility.java
Comment (thobe) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/locking/ActiveLocksListingCompatibility.java]: good point, thank you. I know I did unordered testing elsewhere. will fix.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java
Comment (thobe) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java]: correct.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java
Comment (thobe) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: ok, I'll add that.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/locks/LeaderOnlyLockManager.java
Comment (thobe) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/locks/LeaderOnlyLockManager.java]: ok.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java
Comment (thobe) [reply enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java]: ok.

File: enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java
Comment (thobe) [reply enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java:263]: I implemented it as snapshots since that was easier than streaming on the implementation level. Initially I had streams on the API level but then asked myself "who am I fooling", and changed the API to collections like it was in the implementation.

We need to convert to something with the appropriate public fields, not sure we want to have the internal API use such style of object, but maybe we do. Or perhaps we can turn this into some sort of visitor based (internal) API then have that directly construct the stream of the appropriate objects. That would be a more complicated API, but would imply less allocation.

File: enterprise/kernel/src/test/java/org/neo4j/kernel/enterprise/builtinprocs/ListQueriesProcedureTest.java
Comment (thobe) [reply enterprise/kernel/src/test/java/org/neo4j/kernel/enterprise/builtinprocs/ListQueriesProcedureTest.java]: I used uppercase as a highlight of the central aspect of the test, but I don't mind changing it.

**Comment by thobe** [comment]
Yes, these are Query IDs, not Transaction IDs, since we don't have any IDs fore live transactions. I think it would be better to introduce a Transaction ID, but that was deemed out of scope for the feature request that prompted this work.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java
Comment (thobe) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java]: ... and also as a way to keep these string constants encapsulated.

**Comment by chrisvest** [comment]
Okay. Would it make sense to name the procedure `listActiveLocksForQuery` so we have naming room for a future `listActiveLocks` that takes transaction ids, in case we want to add that later?

**Comment by chrisvest** [comment]
Alternatively, if we add transaction ids later, I suppose the `dbms.listQueries()` could start returning those instead and we can keep the names and queries would keep working?

**Comment by thobe** [comment]
if we add transaction ids later, `dbms.listQueries()` would return _both_ the transaction id and the query id. What we could do though is have the `listActiveLocks` procedure look at the input and do one thing if it is a transaction id, and another thing if it's a query id...

File: enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java
Comment (thobe) [reply enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java:263]: With it now changed to Steams a few places became simpler and could avoid allocation / copying.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java]: Ouch, that's a good point. HashMap is not safe for reads concurrent with writes, because a rehash will reverse the linked list collision buckets, which a concurrent reader than observe as a cycle in the chain, and thus end up in an infinite loop. `ConcurrentHashMap` has slower writes and uses more memory, but I think we probably need it for the lock maps. The maps that segregate by resource type could be turned into arrays and indexed by the resource types ordinal. The probability of seeing an old value in these arrays are probably negligible. It's unfortunate that there's no `AtomicReferenceArrayFieldUpdater`.

**Review by chrisvest — CHANGES_REQUESTED** [review]
I had one comment about the use of `HashMap`s and thread safety in `SlaveLocksClient`. It also needs a rebase, which might also fix the failing test. Otherwise it looks good.

**Comment by thobe** [comment]
@chrisvest yes, that was what I thought would be the right way to do it. I added those TODO notes so that someone would spot it in review and weigh in with another opinion on how to do that. Since we both have the same idea for how to structure it, I'll go ahead and make that change.

**Comment by thobe** [comment]
@chrisvest the rebasing I was taking care of while you were writing your comment 😉

**Comment by thobe** [comment]
@chrisvest fixed.

In a much simpler way. Since the local `Locks.Client` has all the lock information we need (except maybe re-entry counts), and is capable of supporting this API, we can just delegate to that and avoid any concurrency concern in `SlaveLocksClient`.

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
Beautiful.

**Pull request merged**
